### PR TITLE
feat(featureflags): forbid users from saving invalid payloads

### PR DESCRIPTION
**Pull request by dmarticus**
## Problem

Context: https://posthog.slack.com/archives/C034XD440RK/p1727208197973159?thread_ts=1727184377.090709&cid=C034XD440RK

## Changes

Loom: https://www.loom.com/share/0deb70f2c9d34cb5a92ee081179201b3

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

- Unit test backend
- Manual test on the frontend

### CODE REVIEW

**Review by Phanatic — APPROVED** [review]

File: posthog/api/test/test_early_access_feature.py [hunk 1]
```diff
@@ -142,7 +142,7 @@ def test_we_dont_delete_existing_flag_information_when_creating_early_access_fea
                         "rollout_percentage": None,
                     }
                 ],
-                "payloads": {"true": "Hick bondoogling? ????"},
+                "payloads": {"true": '"Hick bondoogling? ????"'},
```
Comment (neilkakkar) [inline posthog/api/test/test_early_access_feature.py:145]: curious, why does this payload turn into a double quoted string? The previous version ought to work too or wouldn't we end up borking just string payloads?

Comment (dmarticus) [reply posthog/api/test/test_early_access_feature.py:145]: Okay, so this is an insightful comment that goes a bit deeper than I expected (although maybe you knew that already and are prompting me to think further).

TL;DR – this will only bork people who are trying to create/update feature flags with string payloads that aren't wrapped in `"` characters – which was my intention: I wanted to make it impossible for folks to put in payloads that can't be JSON parsed.  BUT, existing payloads will still be parsed fine. 

Here's the greater context:

1. right now, every payload that contains a string _with_ spaces in it (e.g., putting in literally `Hick bondoogling? ????`, will be stringified on the frontend through the JSONEditor component and saved in the DB with these double quotes.  So in actuality, this payload that was in the test was never the actual payload that we sent to the backend (I can't share the metabase query with you bc I don't have perms, but running this query
```sql
SELECT filters->'payloads'
FROM posthog_featureflag
WHERE filters->'payloads' IS NOT NULL
  AND jsonb_typeof(filters->'payloads') = 'object'
  AND filters->'payloads' != '{}';
```
will show this).  So in this case, the payload is sent to the DB like this `{"true": "\"Hick bondoogling? ????\""}`

2. That said, the one edge case to this behavior is folks who put in strings with no spaces into the payload, since `JSON.stringify("string")` returns `"string"`.  This means that for spaceless payloads that were entered into the UI form field without quote characters, they're saved into the DB like this `{"true": "nospacepayload"}`.  this means that if a user tried to create a payload that is a string without quotes, we'll fail to save the flag.  That said, this is the behavior I am intending to prevent, since I don't want people saving unparseable JSON as the flag payloads.

3. However, this wasn't a problem for most folks, because in our SDKs (e.g. javascript), we get the payload like this

```
    getFeatureFlagPayload(key: string): JsonType {
        const payload = this.featureFlags.getFeatureFlagPayload(key)
        try {
            return JSON.parse(payload as any)
        } catch {
            return payload
        }
    }
```

`JSON.parse("string")` fails, (but JSON.parse("\"string\"") returns `"string"`), but since we would just return the raw payload if it couldn't be parsed, that was fine for most cases.  It did mean that we broke on things like the situation in [this bug](https://posthog.slack.com/archives/C034XD440RK/p1727184377090709).  

So, what I wanted to do here was make it more explicit for users to ensure that their payloads could be parsed as JSON.  It didn't regress any existing payload parsing behavior for folks that already have technically invalid JSON payloads, but it _might_ cause some confusion if folks go to edit their existing flags (which may have payloads that they want to be strings, but that aren't closed in `"` characters).  I wanted to make that change more explicit, hence this change.  

Does that make sense from your perspective?

Comment (neilkakkar) [reply posthog/api/test/test_early_access_feature.py:145]: nice, makes sense, thanks, the breaking on edit is intentional 👍
